src/mr_allocator.rs

```diff
@@ -1,60 +1,592 @@
+    let errno = unsafe {
+        tikv_jemalloc_sys::mallctl(
+            key.as_ptr().cast(),
+            aid_ptr.cast(),
+            &mut aid_len,
+            ptr::null_mut(),
+            0,
+        )
+    };
+    if errno != 0_i32 {
+        error!("set extent hooks failed");
+        return Err(io::Error::from_raw_os_error(errno));
+    }
+    debug!("create arena success aid : {}", aid);
+    Ok(aid)
+}
+
+/// Create arena and init statics
+fn init_je_statics(pd: Arc<ProtectionDomain>) -> io::Result<u32> {
+    let ind = create_arena()?;
+    if ARENA_PD_MAP.insert(ind, pd) {
+        return Err(io::Error::new(
+            io::ErrorKind::Other,
+            "insert ARENA_PD_MAP failed",
+        ));
+    }
+    set_extent_hooks(ind)?;
+    Ok(ind)
+}
+
+/// Custom extent alloc hook enable jemalloc manage rdma memory region
+#[allow(clippy::expect_used)]
+unsafe extern "C" fn extent_alloc_hook(
+    extent_hooks: *mut extent_hooks_t,
+    new_addr: *mut c_void,
+    size: usize,
+    alignment: usize,
+    zero: *mut i32,
+    commit: *mut i32,
+    arena_ind: u32,
+) -> *mut c_void {
+    let origin_alloc = (*ORIGIN_HOOKS)
+        .alloc
+        .expect("can not get default alloc hook");
+    let addr = origin_alloc(
+        extent_hooks,
+        new_addr,
+        size,
+        alignment,
+        zero,
+        commit,
+        arena_ind,
+    );
+    register_extent_mr_default(addr, size, arena_ind).map_or_else(
+        || {
+            error!("register_extent_mr failed");
+            ptr::null_mut()
+        },
+        |item| {
+            debug!("alloc item {:?} lkey {}", &item, item.raw_mr.lkey());
+            insert_item(item);
+            addr
+        },
+    )
+}
+
+/// Custom extent dalloc hook enable jemalloc manage rdma memory region
+#[allow(clippy::as_conversions)]
+#[allow(clippy::expect_used)]
+unsafe extern "C" fn extent_dalloc_hook(
+    extent_hooks: *mut extent_hooks_t,
+    addr: *mut c_void,
+    size: usize,
+    committed: i32,
+    arena_ind: u32,
+) -> i32 {
+    debug!("dalloc addr {}, size {}", addr as usize, size);
+    remove_item(addr);
+    let origin_dalloc = (*ORIGIN_HOOKS)
+        .dalloc
+        .expect("can not get default dalloc hook");
+    origin_dalloc(extent_hooks, addr, size, committed, arena_ind)
+}
+
+/// Custom extent merge hook enable jemalloc manage rdma memory region
+#[allow(clippy::as_conversions)]
+#[allow(clippy::expect_used)]
+unsafe extern "C" fn extent_merge_hook(
+    extent_hooks: *mut extent_hooks_t,
+    addr_a: *mut c_void,
+    size_a: usize,
+    addr_b: *mut c_void,
+    size_b: usize,
+    committed: i32,
+    arena_ind: u32,
+) -> i32 {
+    debug!(
+        "merge addr_a {}, size_a {}; addr_b {}, size_b {}",
+        addr_a as usize, size_a, addr_b as usize, size_b
+    );
+    let origin_merge = (*ORIGIN_HOOKS)
+        .merge
+        .expect("can not get default merge hook");
+    let err = origin_merge(
```

Comment:
we should check merge boundry and then call `origin_merge`.

Reply:
Je will guarantee only unuesd and adjacent extents are merageable.
Do we need to check it again?

Reply:
We should make sure that cross-mr extents are not merged.